### ISSUE TITLE
loading optee on at91 sama5d2

### ISSUE THREAD

**Issue by kyle-nahrgang** [OP-TEE/optee_os#3964]
**tldr; Is there some obvious mistake I'm making?**

I am using the at91 buildroot and trying to configure optee for my board. Currently, we are booting from a FIT image on a nand device, which is loaded through ubifs. We mount the ubifs, call ubifsload on the FIT, and then call bootm on the address we loaded the fit into. 

The board has the trust zone memory starting at 0x30000000. 

I have attempted three different approaches which all fail in different ways:

### 1.  Loading the existing FIT image (without optee), with optee stored separately in flash
Loading this fit, which boots just fine as is, to 0x20000000 using `ubifsload 0x20000000 /boot/linux.fit`
```
    images {
        kernel@1 {
            data = /incbin/("zImage");
            type = "kernel";
            arch = "arm";
            os = "linux";
            compression = "none";
            load = <0x21000000>;
            entry = <0x21000000>;
        };
        fdt@1 {
            data = /incbin/("linux_device_tree.dtb");
            type = "flat_dt";
            arch = "arm";
            compression = "none";
            fdt-version = <1>;
        };
    };
    configurations {
        default = "conf@1";
        conf@1 {
            description = "Kernel + device tree";
            kernel = "kernel@1";
            fdt = "fdt@1";
        };
    };
```

I call `bootm start 0x2000000`, `bootm loados`, `bootm ramdisk`, `bootm fdt`, and `bootm prep`, which says that the FDT has been loaded to 0x2fb1e000. 

I then load the optee image I created using `mkimage -A arm -T kernel -O tee -C none -d tee.bin uTee.optee` from nand flash to 0x2fffffa4 and try to boot, which hangs at the "starting kernel" message:
```
=> bootm 2fffffa4 - 2fb1e000
*  kernel: cmdline image address = 0x2fffffa4
## Booting kernel from Legacy Image at 2fffffa4 ...
   Image Name:  
   Image Type:   ARM Trusted Execution Environment Kernel Image (uncompressed)
   Data Size:    303900 Bytes = 296.8 KiB
   Load Address: 2fffffe4
   Entry Point:  30000000
   Verifying Checksum ... OK
   XIP Kernel Image ... OK
## initrd_high = 0x30000000, copy_to_ram = 1
   ramdisk load start = 0x00000000, ramdisk load end = 0x00000000
## Skipping init Ramdisk
## No init Ramdisk
   ramdisk start = 0x00000000, ramdisk end = 0x00000000
Wrong FIT format: no description
## Flattened Device Tree blob at 2fb1e000
   Booting using the fdt blob at 0x2fb1e000
## FIT configuration was not specified
   Loading Device Tree to 2fb2e000, end 2fb3bfff ... OK
## Skipping init Ramdisk
## No init Ramdisk
   ramdisk start = 0x00000000, ramdisk end = 0x00000000
Wrong FIT format: no description
## Flattened Device Tree blob at 2fb1e000
   Booting using the fdt blob at 0x2fb1e000
## FIT configuration was not specified
 
Starting kernel ...
```

### 2.    Loading optee as a "loadable" within the fit"
I tried both using that same image generated with mkimage in the first path, as well as the raw tee.bin. Neither have made it very far. 

If I add the optee image to the fit:
```
optee@1 {
            data = /incbin/("uTee.optee");
            type = "kernel";
            arch = "arm";
            os = "tee";
            compression = "none";
            load = <0x2fffffa4>;
            entry = <0x30000000>;
            hash@1 {
                algo = "sha256";
            };
        };
```

Then add the `loadables = "optee@1";` line to `conf@1` shown above, then I consistently get a "loadable corrupt or invalid" error. Turning on some debug messages in uboot says that it couldn't find the parent "/images" for optee@1. 

### 3.    Loading optee as a kernel, with the actual kernel as a loadable
If I use that same image in attempt 2, but change `conf@1` to use `optee@1` as the kernel, and `kernel@1` as the loadable, then I get `No Trusted Execution Environment ARM Kernel Image Image`. 

I am not sure where to look at this point. I feel like attempt 1 got the furthest, but is the least desirable implementation, and attempt 3 didn't go anywhere but is probably the most desirable. 

Is there some obvious mistake I'm making? I have gone through the u-boot configuration a few times and have all op-tee related things enabled (except those in the SPL config, since my board doesn't appear to support it).

**Comment by sckulkarni246** [comment]
Hi Kyle,

I am currently using the below strategy to load optee on a sama5d2 xplained board:

1) at91bootstrap loads u-boot
2) u-boot copies dtb to 0x21000000, zImage to 0x23000000, uTee to 0x22000000
3) bootm 0x22000000 - 0x21000000

The optee is built with NS ENTRY address as 0x22000000 (load address of zImage in step 2 above).

The boot flow is as below:
at91bootstrap -> u-boot -> optee -> kernel -> rootfs

I don't use FIT. Let me know if you need the image.

Regards,
Shashank

**Comment by kyle-nahrgang** [comment]
@sckulkarni246 I am not using the sama5d2 xplained board, so I'm not sure if there are some little differences. The TZ memory is up above 0x30000000 for me so I would've thought that uTee would need to be loaded up closer to that? In u-boot, what are your TZ RAM settings? Mine look like this:
```
[*] Support OPTEE images
(0x2FFFFFE4) OPTEE load address
(0xC00000) Amount of Trust-Zone RAM for the OPTEE image   │ │                                                                                              (0x30000000) Base address of Trust-Zone RAM for the OPTEE image
```

Nevertheless, without using the fit. I compiled optee-os with CFG_NS_ENTRY_ADDR=0x23000000 (I believe you had a typo in your comment), load the FDT to 0x21000000, the tee to 0x22000000, and the zImage to 0x23000000. 

Calling `bootm 0x22000000 - 0x21000000` still hangs at "Starting kernel..." :
```
=> bootm 0x22000000 - 0x21000000
## Booting kernel from Legacy Image at 22000000 ...
   Image Name:   
   Image Type:   ARM Trusted Execution Environment Kernel Image (uncompressed)
   Data Size:    303900 Bytes = 296.8 KiB
   Load Address: 2fffffe4
   Entry Point:  30000000
   Verifying Checksum ... OK
   Loading Kernel Image ... OK
## Flattened Device Tree blob at 21000000
   Booting using the fdt blob at 0x21000000
   Loading Device Tree to 2fb2f000, end 2fb3c08a ... OK
## Flattened Device Tree blob at 21000000
   Booting using the fdt blob at 0x21000000

Starting kernel ...
```

Moving the uTee image up higher gives slightly different output but still hangs:
```
=> bootm 0x2fffffa4 - 0x21000000
## Booting kernel from Legacy Image at 2fffffa4 ...
   Image Name:   
   Image Type:   ARM Trusted Execution Environment Kernel Image (uncompressed)
   Data Size:    303900 Bytes = 296.8 KiB
   Load Address: 2fffffe4
   Entry Point:  30000000
   Verifying Checksum ... OK
   XIP Kernel Image ... OK
## Flattened Device Tree blob at 21000000
   Booting using the fdt blob at 0x21000000
   Loading Device Tree to 2fb2f000, end 2fb3c08a ... OK
## Flattened Device Tree blob at 21000000
   Booting using the fdt blob at 0x21000000

Starting kernel ...
```

**Comment by kyle-nahrgang** [comment]
Documentation about the memory ranges: https://github.com/OP-TEE/optee_os/blob/master/core/arch/arm/plat-sam/platform_config.h

**Comment by kyle-nahrgang** [comment]
I realized that optee_os was not using the correct debug console so I managed to actually get it to boot with some output.

```
Starting kernel ...

D/TC:0 0 add_phys_mem:536 VCORE_UNPG_RX_PA type TEE_RAM_RX 0x30000000 size 0x0004d000
D/TC:0 0 add_phys_mem:536 VCORE_UNPG_RW_PA type TEE_RAM_RW 0x3004d000 size 0x000b3000
D/TC:0 0 add_phys_mem:536 TA_RAM_START type TA_RAM 0x30100000 size 0x00700000
D/TC:0 0 add_phys_mem:536 TEE_SHMEM_START type NSEC_SHM 0x30800000 size 0x00400000
D/TC:0 0 add_phys_mem:536 CONSOLE_UART_BASE type IO_SEC 0xfc000000 size 0x00200000
D/TC:0 0 add_phys_mem:536 PL310_BASE type IO_SEC 0x00a00000 size 0x00100000
D/TC:0 0 add_phys_mem:536 SFR_BASE type IO_SEC 0xf8000000 size 0x00200000
D/TC:0 0 add_phys_mem:536 AT91C_BASE_MATRIX32 type IO_SEC 0xfc000000 size 0x00200000
D/TC:0 0 add_phys_mem:549 Physical mem map overlaps 0xfc000000
D/TC:0 0 add_phys_mem:536 AT91C_BASE_MATRIX64 type IO_SEC 0xf0000000 size 0x00200000
D/TC:0 0 verify_special_mem_areas:474 No NSEC DDR memory area defined
D/TC:0 0 add_va_space:575 type RES_VASPACE size 0x00a00000
D/TC:0 0 add_va_space:575 type SHM_VASPACE size 0x02000000
D/TC:0 0 dump_mmap_table:708 type TEE_RAM_RX   va 0x30000000..0x3004cfff pa 0x30000000..0x3004cfff size 0x0004d000 (smallpg)
D/TC:0 0 dump_mmap_table:708 type TEE_RAM_RW   va 0x3004d000..0x300fffff pa 0x3004d000..0x300fffff size 0x000b3000 (smallpg)
D/TC:0 0 dump_mmap_table:708 type SHM_VASPACE  va 0x30100000..0x320fffff pa 0x00000000..0x01ffffff size 0x02000000 (pgdir)
D/TC:0 0 dump_mmap_table:708 type IO_SEC       va 0x32100000..0x322fffff pa 0xfc000000..0xfc1fffff size 0x00200000 (pgdir)
D/TC:0 0 dump_mmap_table:708 type RES_VASPACE  va 0x32300000..0x32cfffff pa 0x00000000..0x009fffff size 0x00a00000 (pgdir)
D/TC:0 0 dump_mmap_table:708 type TA_RAM       va 0x32d00000..0x333fffff pa 0x30100000..0x307fffff size 0x00700000 (pgdir)
D/TC:0 0 dump_mmap_table:708 type NSEC_SHM     va 0x33400000..0x337fffff pa 0x30800000..0x30bfffff size 0x00400000 (pgdir)
D/TC:0 0 dump_mmap_table:708 type IO_SEC       va 0x33800000..0x338fffff pa 0x00a00000..0x00afffff size 0x00100000 (pgdir)
D/TC:0 0 dump_mmap_table:708 type IO_SEC       va 0x33900000..0x33afffff pa 0xf0000000..0xf01fffff size 0x00200000 (pgdir)
D/TC:0 0 dump_mmap_table:708 type IO_SEC       va 0x33b00000..0x33cfffff pa 0xf8000000..0xf81fffff size 0x00200000 (pgdir)
D/TC:0 0 core_mmu_alloc_l2:238 L2 table used: 1/4
I/TC: 
D/TC:0 0 init_canaries:166 #Stack canaries for stack_tmp[0] with top at 0x30076838
D/TC:0 0 init_canaries:166 watch *0x3007683c
D/TC:0 0 init_canaries:167 #Stack canaries for stack_abt[0] with top at 0x30077078
D/TC:0 0 init_canaries:167 watch *0x3007707c
D/TC:0 0 init_canaries:169 #Stack canaries for stack_thread[0] with top at 0x300790b8
D/TC:0 0 init_canaries:169 watch *0x300790bc
D/TC:0 0 init_canaries:169 #Stack canaries for stack_thread[1] with top at 0x3007b0f8
D/TC:0 0 init_canaries:169 watch *0x3007b0fc
I/TC: OP-TEE version: linux4sam_6.1 #4 Tue Jul  7 18:49:29 UTC 2020 arm
D/TC:0 0 check_ta_store:533 TA store: "Secure Storage TA"
D/TC:0 0 check_ta_store:533 TA store: "REE"
D/TC:0 0 mobj_mapped_shm_init:709 Shared memory address range: 30100000, 32100000
E/TC:0 0 plat_rng_init:354 Warning: seeding RNG with zeroes
I/TC: Initialized
D/TC:0 0 init_primary_helper:1033 Primary CPU switching to normal world boot
Booting Linux on physical CPU 0x0
```

This is progress, but when I `modprobe optee` the whole device hangs.

**Comment by sckulkarni246** [comment]
Hi @kyle-nahrgang ,

Have you made the changes to the device tree to reserve memory for OPTEE?
Something like attached. I have added these to the end of my device tree.

![dt-optee](https://user-images.githubusercontent.com/1713224/86890391-e1cf6f80-c11a-11ea-8771-72f50c106af3.png)

Why not build TEE support into the kernel rather than making it loadable?

Regards,
Shashank

**Comment by kyle-nahrgang** [comment]
Thank you for your help here, it's been definitely helping me make some forward progress. 

When I attempt to reserve that memory in the device tree, uboot throws an error. Here is the output after adding some extra debugging:
```
lmb_add_region: regions overlap
lmb_dump_all:
    memory.cnt             = 0x1
    memory.size            = 0x0
    memory.reg[0x0].base   = 0x20000000
                   .size   = 0x10000000
    reserved.cnt           = 0x3
    reserved.size          = 0x0
    reserved.reg[0x0].base = 0x21000000
                     .size = 0x3e6550
    reserved.reg[0x1].base = 0x2fb41000
                     .size = 0xd10e
    reserved.reg[0x2].base = 0x2fb4ec60
                     .size = 0x10b13a0
ERROR: reserving fdt memory region failed (addr=30000000 size=c00000)
```

It looks like the last region there is actually going up to 0x30C00000 which makes me think it's being reserved by SOMETHING. I have not found that block referenced in any part of the DTS I am adding, or sama5d2.dtsi. 

Before, I would get a kernel panic when trying to load the optee driver because arm_smccc_smc was null. Now when that function is called it just freezes up, though.

**Comment by sckulkarni246** [comment]
Hi @kyle-nahrgang - do you have a SAMA5D2 Xplained handy? Maybe we can start off by seeing what configs are different compared to my working image for that board.
My working image is built on top of a linux4sam image. Let me know  if you have a board handy - I will share the SD card image and the config with you.

Regards,
Shashank

**Comment by kyle-nahrgang** [comment]
We are actually using Atmel SAMA5D2 PTC EK, not the Xplained board, and even that has some customizations so the images wouldn't be particularly helpful. 

For at91bootstrap, we are using sama5d2_ptc_eknf_uboot_defconfig.
In our u-boot config, I am making sure we set:

- Enable support for the legacy image format  
- Support OPTEE images
- OPTEE load address to 0x30000000
-  Amount of Trust-Zone RAM for the OPTEE image 0xC00000 
                                                                                                                                             │ │  
  │ │                                                                                              (0x30000000) OPTEE load address                                                                                                                                      │ │  
  │ │                                                                                              (0x800000) Amount of Trust-Zone RAM for the OPTEE image                                                                                                              │ │  
  │ │                                                                                              (0x30000000) Base address of Trust-Zone RAM for the OPTEE image                                                                                                      │ │  
  │ │                                                                                              [*] Support OPTEE bootm command

**Comment by sckulkarni246** [comment]
Hi @kyle-nahrgang - sorry for the delayed response. Was caught up with some personal stuff over the last couple of days.

The process I followed to generate an image is as follows:

1) Generated a microchip base image using yocto (did microchip-headless-image and microchip-graphics-image)
2) No changes done to at91bootstrap
3) No changes done to u-boot
4) Added TEE support to kernel and modified device tree as hinted above
5) Added optee-client binaries to the rootfs
6) [OPTIONAL] Added xtest application to the rootfs 

This is a link to the video with the image booting correctly and also running the xtest.
[demo link](https://microchiptechnology-my.sharepoint.com/:v:/g/personal/shashank_k_microchip_com/EZsdwN2FIWlJmTYIF4mN038Bt--OII-30Xl-uMjh297mZA?e=9LSoz3) 

You should be able to follow the steps exactly as above and build an image. I have recently helped someone do this on their custom board (again eMMC based board).
Let me know how it goes!

Regards,
Shashank

**Comment by kyle-nahrgang** [comment]
Sorry, I haven't been able to focus on this much in the last week. 

I see that in your demo video you were also getting the error from UBoot when trying to reserve the memory for optee. I think I may have seen that error and never actually fully booted linux. I will attempt that in the next day or two and update again.

**Comment by kyle-nahrgang** [comment]
I confirmed that reserving memory still does not change the error during boot. I did notice that these lines are different in your boot process:
```
D/TC:0 0 add_phys_mem:536 VCORE_UNPG_RX_PA type TEE_RAM_RX 0x30000000 size 0x0004d000
D/TC:0 0 add_phys_mem:536 VCORE_UNPG_RW_PA type TEE_RAM_RW 0x3004d000 size 0x000b3000
```

Does your board have secure boot enabled, where at91bootstrap must be encrypted?
